IO Classes

pblivingston opened the pull request:
## What

Centralizes controls common to both Strip and Bus.

### Adds

- `IOControl` class with `mute`, `gain`, `label`, `FadeTo()`, `FadeBy()`
- `IOLevels` class with `Getter()` and hidden `Convert()`

### Changes

- `Eq` class is now `IOEq`
- `Device` class is now `IODevice`

## Testing summary

Pester tests pass for all kinds

pblivingston committed:
Create io.ps1

- IOControl
- IOLevels
- IOEq
- IODevice

pblivingston committed:
implement io classes

prelim pester tests for potato pass

pblivingston committed:
stripknob

centralizes threshold, knob
- stripcomp
- stripgate
- stripdenoiser

prelim pester tests for potato pass

pblivingston committed:
io classes, stripknob to CHANGELOG

pester tests pass for all kinds

pblivingston commented:
With the changes from #33, `IODevice` becomes only a few lines. Although there should be benefits to centralizing behavior in io.ps1 like this, another option would be to simply revert the decoupling of `Device` from `Strip` and `Bus` as the class will be very simple in each, and only leave eq.ps1.

Though I think centralizing things might be better for maintenance of equivalent behavior and organizes it in a way that I personally find sensible and readable, keeping the simpler parts separate might be more readable to others.

Let me know your thoughts!

onyx-and-iris commented:
Generally I like this change, just one thing, I think if we're going to introduce a class `StripKnob` as a base class for the physical strip knobs then Audibility should subclass it. I see why you didn't do that because you moved `AddFloatMembers -PARAMS @('threshold')` into StripKnob but Audibility is a kind of strip knob. So in order to keep the class hierarchy most meaningful I suggest moving `AddFloatMembers -PARAMS @('threshold')` back into the comp/gate/denoiser derived classes. Yes this means a little repetition but it's explicit and ensures Audibility is a member of that hierarchy, which is where I think it belongs.

Other than that I'm happy with this PR.

pblivingston commented:
Okay, sounds good!

pblivingston committed:
knob threshold, audibility

move threshold back to comp/gate/denoiser so audibility can be derived from StripKnob

pester tests for safety pass

pblivingston commented:
That's all set!
